### PR TITLE
Enrich the TraceReplayEvent with initial point

### PR DESCRIPTION
**Pull request by Jasagredo**
This will allow to calculate percentages of completion on ledger replay
tracing logic on the node.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 1]
```diff
@@ -158,6 +164,32 @@ data InitLog blk =
...
+--
+-- Between the tip of the immutable DB and the point of the starting block,
+-- the node could (if it so desired) easily compute a "percentage complete".
+decorateReplayTracerWithGoal
+  :: Point blk -- ^ Tip of the ImmutableDB
+  -> Tracer m (TraceLedgerReplayEvent blk)
+  -> Tracer m (TraceReplayEvent blk (Point blk, ()))
+decorateReplayTracerWithGoal immTip = contramap $ fmap (\(x, _) -> (x, immTip))
+
+decorateReplayTracerWithStart
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:187]: Please change this to 

```
  :: ReplayStart blk
  -> Tracer m (TraceReplayEvent blk (ReplayStart blk))
  -> Tracer m (TraceReplayEvent blk ()
```

or -- if we go with my new suggestion -- to:

```
  :: ReplayStart blk
  -> Tracer m (ReplayGoal blk -> TraceReplayEvent blk)
  -> Tracer m (ReplayStart blk -> ReplayGoal blk -> TraceReplayEvent blk)
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 2]
```diff
@@ -158,6 +164,32 @@ data InitLog blk =
...
+--
+-- The @replayTo@ parameter is instantiated with the 'Point' of
+-- the tip of the ImmutableDB.
+type TraceLedgerReplayEvent blk = TraceReplayEvent blk (Point blk, Point blk)
+
+-- | Add the tip of the Immutable DB to the trace event
+--
+-- Between the tip of the immutable DB and the point of the starting block,
+-- the node could (if it so desired) easily compute a "percentage complete".
+decorateReplayTracerWithGoal
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:181]: Please change this to 

```
  :: ReplayGoal blk
  -> Tracer m (TraceReplayEvent blk (ReplayStart blk, ReplayGoal blk))
  -> Tracer m (TraceReplayEvent blk (ReplayStart blk)
```

or -- if we go with my new suggestion -- to:

```
  :: ReplayGoal blk
  -> Tracer m (TraceReplayEvent blk)
  -> Tracer m (ReplayGoal blk -> TraceReplayEvent blk)
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/LgrDB.hs [hunk 3]
```diff
@@ -175,7 +175,7 @@ openDB :: forall m blk.
           )
        => LgrDbArgs Identity m blk
        -- ^ Stateless initializaton arguments
-       -> Tracer m (TraceReplayEvent blk ())
+       -> Tracer m (TraceReplayEvent blk (Point blk, ()))
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/LgrDB.hs:178]: Please change this argument type to `TraceReplayEvent blk (ReplayStart blk)`

Or to `ReplayGoal blk -> TracerReplayEvent blk`, if we go with my new suggestion.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/LgrDB.hs [hunk 4]
```diff
@@ -236,7 +236,7 @@ initFromDisk
      , HasCallStack
      )
   => LgrDbArgs Identity m blk
-  -> Tracer m (TraceReplayEvent blk ())
+  -> Tracer m (TraceReplayEvent blk (Point blk, ()))
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/LgrDB.hs:239]: Please change this argument type to `TraceReplayEvent blk (ReplayStart blk)`

Or to `ReplayGoal blk -> TracerReplayEvent blk`, if we go with my new suggestion.

**Review by nfrisby — CHANGES_REQUESTED** [review]
I wrote several types in my comments. I _think_ they're right, but I did it in my head, so I could be wrong. I'm pretty sure the types at the `decorate*` functions' comments are right, but I'm less confident about the use cases.

Summary: there are too many `Point`s and `()`s floating around in the current PR. So let's either use `newtypes` to clarify, or else use the new idea I came up with that uses function spaces (see the comment).

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 5]
```diff
@@ -158,6 +164,32 @@ data InitLog blk =
   | InitFailure DiskSnapshot (InitFailure blk) (InitLog blk)
   deriving (Show, Eq, Generic)
 
+{-------------------------------------------------------------------------------
+  TraceReplayEvent decorator
+-------------------------------------------------------------------------------}
+
+-- | 'TraceReplayEvent' instantiated with additional information.
+--
+-- The @replayTo@ parameter is instantiated with the 'Point' of
```
[the remark this answers is not in the document]
Comment (Jasagredo) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs]: I adopted the suggestion and now the incomplete event types are being used.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 6]
```diff
@@ -299,7 +334,7 @@ initStartingWith ::
        , InspectLedger blk
        , HasCallStack
        )
-  => Tracer m (TraceReplayEvent blk ())
+  => Tracer m (TraceReplayEvent blk ((), ()))
```
[the remark this answers is not in the document]
Comment (Jasagredo) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs]: Done

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs (hunk 1 above)
Comment (Jasagredo) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs]: Done

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs (hunk 2 above)
Comment (Jasagredo) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs]: Done

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/LgrDB.hs (hunk 4 above)
Comment (Jasagredo) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/LgrDB.hs]: Done

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/LgrDB.hs (hunk 3 above)
Comment (Jasagredo) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/LgrDB.hs]: Done

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 7]
```diff
@@ -186,7 +218,7 @@ initLedgerDB ::
        , InspectLedger blk
        , HasCallStack
        )
-  => Tracer m (TraceReplayEvent blk ())
+  => Tracer m (TraceReplayEvent blk (Point blk, ()))
```
[the remark this answers is not in the document]
Comment (Jasagredo) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs]: Done

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 8]
```diff
@@ -268,7 +301,7 @@ initFromSnapshot ::
...
```
[the remark this answers is not in the document]
Comment (Jasagredo) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs]: Done

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 9]
```diff
@@ -499,6 +508,23 @@ snapshotFromPath fileName = do
...
+--
+-- Between the tip of the immutable DB and the point of the starting block,
+-- the node could (if it so desired) easily compute a "percentage complete".
+decorateReplayTracerWithGoal
+  :: Point blk -- ^ Tip of the ImmutableDB
+  -> Tracer m (TraceReplayEvent blk)
+  -> Tracer m (ReplayGoal blk -> TraceReplayEvent blk)
+decorateReplayTracerWithGoal immTip = contramap ($ (ReplayGoal immTip))
+
+decorateReplayTracerWithStart
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:522]: This one is missing some Haddock.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 10]
```diff
@@ -508,32 +534,36 @@ data TraceEvent blk
     -- ^ An old or invalid on-disk snapshot was deleted
   deriving (Generic, Eq, Show)
 
+
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:537]: there's an extra newline here

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 11]
```diff
@@ -508,32 +534,36 @@ data TraceEvent blk
...
+  | ReplayedBlock (RealPoint blk) [LedgerEvent blk] (ReplayStart blk) (ReplayGoal blk)
     -- ^ We replayed the given block (reference) on the genesis snapshot
     -- during the initialisation of the LedgerDB.
     --
     -- The @blockInfo@ parameter corresponds replayed block and the @replayTo@
     -- parameter corresponds to the block at the tip of the ImmutableDB, i.e.,
     -- the last block to replay.
   | UpdateLedgerDbTraceEvent (UpdateLedgerDbTraceEvent blk)
-  deriving (Generic, Eq, Show, Functor, Foldable, Traversable)
+  deriving (Generic, Eq, Show)
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:569]: Minor: Can you also remove the `DeriveFunctor`, `DeriveFoldable`, and `DeriveTraversable` LANGUAGE pragmas now?

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 12]
```diff
@@ -508,32 +534,36 @@ data TraceEvent blk
...
 -- | Events traced while replaying blocks against the ledger to bring it up to
 -- date w.r.t. the tip of the ImmutableDB during initialisation. As this
 -- process takes a while, we trace events to inform higher layers of our
 -- progress.
---
--- The @replayTo@ parameter is meant to be filled in by a higher layer,
--- i.e., the ChainDB.
-data TraceReplayEvent blk replayTo
-  = ReplayFromGenesis replayTo
+data TraceReplayEvent blk
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:548]: We don't necessarily need to use the new goal and start newtypes in here; we could just use `Point`. But the newtypes are probably clearer, ultimately. Even though they'll require some (minor and probably more legible!) swizzling by downstream users.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 13]
```diff
@@ -508,32 +534,36 @@ data TraceEvent blk
...
+  = ReplayFromGenesis (ReplayGoal blk)
     -- ^ There were no LedgerDB snapshots on disk, so we're replaying all
     -- blocks starting from Genesis against the initial ledger.
     --
     -- The @replayTo@ parameter corresponds to the block at the tip of the
     -- ImmutableDB, i.e., the last block to replay.
-  | ReplayFromSnapshot DiskSnapshot (RealPoint blk) replayTo
+  | ReplayFromSnapshot DiskSnapshot (RealPoint blk) (ReplayStart blk) (ReplayGoal blk)
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:555]: These three constructors all need to update their Haddock comment to stop mentioning `replayTo` and instead explain the two points.

**Review by nfrisby — COMMENTED** [review]
Nice and clean!

I wrote some minor comments. EG some Haddock is out of date.

Lastly: it might be helpful to mention the simplification (and changes) in the `interface-CHANGELOG.md`. Just a sentence or two explaining the new types that the event's fields have when its being processed by downstream code.

Once you've addressed these, feel free to have Damian/Pawel/me (whoever is online) Approve for merge.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs (hunk 9 above)
Comment (Jasagredo) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:524]: Addressed

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs (hunk 10 above)
Comment (Jasagredo) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs]: Removed

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs (hunk 11 above)
Comment (Jasagredo) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:572]: Addressed

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs (hunk 12 above)
Comment (Jasagredo) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:549]: I would keep them for now. They don't do any harm and we are the ones that will update the downstream deps so we can deal with it :smile:

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs (hunk 13 above)
Comment (Jasagredo) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs]: Updated

**Review by nfrisby — APPROVED** [review]
Approved! Everything I raised was addressed 👍

I think this can be rebased and merged after PR #3537.

**Comment by nfrisby** [comment]
Now that PR #3537 has merged, I did a rebase. There were no conflicts.

**Comment by nfrisby** [comment]
bors merge

**Comment by nfrisby** [comment]
bors retry

**Comment by nfrisby** [comment]
bors merge

**Pull request merged**